File: module-test/src/main/java/com/jojo/design/module_test/room/AppDatabase.java
package com.jojo.design.module_test.room;

import androidx.sqlite.db.SupportSQLiteDatabase;
import androidx.room.Database;
import androidx.room.RoomDatabase;
import androidx.room.migration.Migration;

/**
 * author : JOJO
 * e-mail : 18510829974@163.com
 * date   : 2019/1/4 5:56 PM
 * desc   :
 */
//注解指定了database的表映射实体数据以及版本等信息
@Database(entities = {User.class}, version = 2)
public abstract class AppDatabase extends RoomDatabase {
    //RoomDatabase提供直接访问底层数据库实现，我们通过定义抽象方法返回具体Dao
//然后进行数据库增删该查的实现。
    public abstract UserDao userDao();

    //数据库变动添加Migration  :[数据库操作语句大全(sql)](https://www.cnblogs.com/liuzheng-java/articles/sql.html)
    public static final Migration MIGRATION_1_2 = new Migration(1, 2) {
        @Override
        public void migrate(SupportSQLiteDatabase database) {
            //user中添加一个新字段age
            database.execSQL("ALTER TABLE user "
                    + " ADD COLUMN age INTEGER NOT NULL DEFAULT 0");
        }
    };
}

File: module-mall/src/main/java/com/jojo/design/module_mall/db/AppDataBase.kt
package com.jojo.design.module_mall.db

import android.arch.persistence.room.Database
import android.arch.persistence.room.Room
import android.arch.persistence.room.RoomDatabase
import android.content.Context
import com.jojo.design.module_mall.db.bean.SearchHistoryBean
import com.jojo.design.module_mall.db.dao.HistoryDao

/**
 *    author : JOJO
 *    e-mail : 18510829974@163.com
 *    date   : 2019/1/2 3:17 PM
 *    desc   :
 */
@Database(entities = arrayOf(SearchHistoryBean::class, SearchHistoryBean::class), version = 1)
abstract class AppDataBase constructor(context: Context) : RoomDatabase() {
    private var DB_NAME = "designer.db"
    @Volatile
    private var instance: AppDataBase? = null

//    https://www.jianshu.com/p/cfde3535233d
    @Synchronized
     fun getInstance(context: Context): AppDataBase {
        if (instance == null) {
            instance = create(context)
        }
        return instance!!
    }

   open fun create(context: Context): AppDataBase? = Room.databaseBuilder(context, AppDataBase::class.java, DB_NAME).build()

    abstract fun historyDao(): HistoryDao

}